src/order/basic.lean

```diff
@@ -612,6 +614,37 @@ or_iff_not_imp_left.2 $ λ h,
   ⟨λ a ha₁, le_of_not_gt $ λ ha₂, h ⟨a, ha₁, ha₂⟩,
     λ a ha₂, le_of_not_gt $ λ ha₁, h ⟨a, ha₁, ha₂⟩⟩
 
+/-- Order equipped with a sensible successor function. -/
+class succ_order (α : Type u) [preorder α] :=
+(succ : α → α)
+(lt_iff_succ_le : ∀ a b, a < b ↔ succ a ≤ b)
+(lt_succ_iff_le : ∀ a b, a < succ b ↔ a ≤ b)
+
+open succ_order
+
+lemma lt_iff_succ_le {α : Type u} [preorder α] [succ_order α] {a b : α} :
+  a < b ↔ succ a ≤ b :=
+lt_iff_succ_le a b
+
+lemma lt_succ_iff_le {α : Type u} [preorder α] [succ_order α] {a b : α} :
+  a < succ b ↔ a ≤ b :=
+lt_succ_iff_le a b
+
+/-- Class stating that `∀ a b, a < b ↔ a + 1 ≤ b` and `∀ a b, a < b + 1 ↔ a ≤ b`. -/
+class succ_eq_add_one_order (α : Type u) [preorder α] [has_add α] [has_one α] extends
```

Comment:
A note: We should be cautious whenever introducing a typeclass that takes typeclasses as arguments: it can make term sizes grow exponentially. I expect that this case won't be too bad since `preorder`/`has_add`/`has_one` are basically roots of the inheritance diagram, and the alternative (extending rather than parametrizing) would require adding a new typeclass for each class for `nat` or `int` that extends `preorder`, `has_add` and/or `has_one`, which are way too many.

Reply:
I'm actually wondering whether this "syntaxes agree" typeclass is even useful. Can't we simply rely on defeqness of `succ` and `λ a, a + 1` in the relevant cases?